### ISSUE TITLE
`_metadata` is re-read for every partition when using `fastparquet` engine

### ISSUE THREAD

**Issue by ig248** [dask/dask#5842]
## Background
Given a (poorly) partitioned dataset of 15000 partitions of ~3k rows each, with a `_metadata` file, we evaluate e.g.

```python
df = dd.read_parquet(source, engine=engine)  # L1: mostly read _metadata
df.head()  # L2: single task
df.size.compute()  # L3: 15000 tasks
```

with engine set to either `fastparquet` or `pyarrow`. After [this speed-up](https://github.com/dask/dask/pull/5821), L1 takes ~25s with `fastparquet` vs 215s with `pyarrow`.

However, L2 takes ~20s with `fastparquet` vs ~200-300ms with `pyarrow`.

Similarly, L3 takes ~20s **per partition** (`read-parquet-...` task) on the workers - this appears to be largely due to repeated re-reading the metadata file, as confirmed by removing the `_metadata` file after executing L1.

This makes processing the data almost 100x slower than it could be, and leads to `O(n^2)` time in size of data when partitions are small.

**Comment by jakirkham** [comment]
We are probably doing that by default as we can't assume much about a general parquet file. That said, if you do know more about your data, you may consider setting some of the other arguments to `read_parquet` to provide it better info about your dataset. Am curious if this is something you have tried. If not, would you be willing to explore this a bit and report back?

**Comment by ig248** [comment]
@jakirkham thanks, I will definitely explore the issue further. However, I'm fairly sure there's an issue here given that repeated _metadata reads do not occur with pyarrow, and we should be able to get the full benefit of the _metadata file from a single (currently client-side) read. I will update once I had a bit of time to investigate, but it seems desirable that `read_parquet_partition` should only need a the relevant file path and the relevant row group metadata, unpacked from the initial _metadata read

**Comment by jakirkham** [comment]
That sounds possible. I'm hoping we can narrow down a bit more where the issue is coming from with this investigation 🙂

**Comment by mrocklin** [comment]
Sorry for the lack of response here @ig248 and thank you for raising this issue.  I think that the first step to understanding the performance here is probably to construct a minimal example.  Probably this would generate a small dataset with lots of partitions (hopefully something easily tunable).  This is something that folks could probably run locally relatively easily and profile to see where the hotspots are.  

It has been a while since you raised this (my apologies) but if you're still interested in this and have time to help, that would be welcome.

**Comment by ig248** [comment]
@mrocklin definitely, I will get back to this issue once I'm a bit less buried

**Comment by ig248** [comment]
@mrocklin I have added a reproducible example and profiler output showing what code path is causing the issue.

**Comment by ig248** [comment]
Looks like @rjzamora may have intended for `_metadata` to be re-read in [this commit](https://github.com/dask/dask/commit/6dfc8e9ab8c27cdb18847ed92349aa817640dcd3) - @rjzamora ?

**Comment by rjzamora** [comment]
> Looks like @rjzamora may have intended for _metadata to be re-read in this commit - @rjzamora ?

Thanks for raising @ig248 - Sorry about this!  You are correct that there is some pretty inefficient code in `FastParquetEngine`.  The origin of the change you highlighted was that the size of the task graph was becoming a serious issue for some people.   Background: Fastparquet requires one to produce a `ParquetFile` object to read in an individual row-group.  We originally read in `"_metadata"` once to create this object (as you would expect), and then passed the  object to the workers throught the graph.  Unfortunately, the size of the `ParquetFile` object was resulting in extreme run times at scale.  So, as a "quick fix", we removed the `pf` object from the graph, and required each individual worker to read `"_metadata"` (when the shared file was available).

In hindsight, it seems pretty obvious that parsing `"_metadata"` before reading each row-group is a terrible idea. I'll admit that I am not really a user of `FastParquetEngine`, so I never did much/any real performance testing.  With that said,  there are probably ways we can avoid the new problem I created - I'll think on this for a bit today.  I'm also open to suggestions from others, like yourself and @martindurant :)

**Comment by ig248** [comment]
@rjzamora I agree with reducing graph size, and I think `fast_metadata` is a good approach. Passing a single string instead of the full `ParquetFile` makes sense, however I'd suggest some ways to improve performance:
(1) avoid call to `_analyze_paths` by passing the full path to `_metadata` instead of path to root folder,
(2) ideally avoid call to `_determine_pf_parts` from `read_partition` altogether and passing string straight to `ParquetFile()`
(3) Finally, to address the bigger issue of re-reading `_metadata`, I would consider creating a surgically modified ParquetFile with all but the relevant row group removed? This way we may be able to find a balance between graph size and read speed.

Maybe the `fast_metadata` could be a kwarg, or chosen based on the ratio of number of row groups to number of rows per row group, to control whether we use (3) or (1) above.
